cogs/strawpoll.py

```diff
@@ -0,0 +1,292 @@
+                value=self._bar_creator(data, option))
+        time_left = round(self.poll_length-(time()-self.start_time))
+        if (time_left > 0):
+            embed.set_footer(text="{} sec{} left".format(
+                str(time_left), '' if (time_left == 1) else 's'))
+        else: 
+            embed.set_footer(text="as of {}".format(
+                strftime('%A %B %-m, %Y @ %I:%M:%S%p ')))
+        await self.bot.edit_message(self.message, embed=embed)
+
+class Strawpoll:
+    """Create, link, and update live results for a Strawpoll within Discord"""
+    
+    def __init__(self, bot, settings_path):
+        self.bot = bot
+        self.settings_path = settings_path
+        self.poll_sessions = []
+        self.poll_session_tasks = {}
+        try:
+            if not dataIO.is_valid_json(settings_path):
+                raise FileNotFoundError("JSON file not valid")
+            settings = dataIO.load_json(settings_path)
+            self.settings = defaultdict(dict, settings)
+        except FileNotFoundError:
+            self.settings = {}
+            if not os.path.exists(os.path.dirname(settings_path)):
+                print("Creating strawpoll data folder...")
+                os.makedirs(os.path.dirname(settings_path))
+            dataIO.save_json(settings_path, self.settings)
+
+    def _new_server_settings(self, server_id):
+        default_settings = {
+            'refresh_emoji': '🔄',
+            'poll_length': 60.0,
+            'poll_react_time': 1800.0, # 30 minutes
+            'bar_length': 30
+        }
+        self.settings[server_id] = default_settings
+        dataIO.save_json(self.settings_path, self.settings)
+
+    def _check_new_poll_tasks(self):
+        for poll in self.poll_sessions:
+            if poll not in self.poll_session_tasks:
+                self.poll_session_tasks[poll] = asyncio.ensure_future(
+                    self._check_reacts(poll))
+
+    def _cancel_expired_poll_tasks(self, current_time):
+        for poll in self.poll_session_tasks.keys():
+            server_id = poll.message.server.id
+            poll_react_time = self.settings[server_id]['poll_react_time']
+            if current_time-poll.start_time > poll_react_time:
+                self.poll_session_tasks[poll].cancel()
+ 
+    def _delete_expired_poll_sessions(self, current_time):
+        trimmed_poll_sessions = []
+        for poll in self.poll_sessions:
+            server_id = poll.message.server.id
+            poll_react_time = self.settings[server_id]['poll_react_time']
+            if current_time-poll.start_time < poll_react_time:
+                trimmed_poll_sessions.append(poll)
+        self.poll_sessions = trimmed_poll_sessions
+
+    def _remove_done_poll_tasks(self):
+        self.poll_session_tasks = { 
+            poll:task for poll,task in self.poll_session_tasks.items() 
+            if not task.done()}
+
+    async def check_polls(self):
+        while self is self.bot.get_cog("Strawpoll"):
```

Comment:
I've not seen this syntax before. What is this for? Maybe "while cog is loaded"?

Reply:
stream_checker uses this syntax in streams.py